dotCMS/src/main/java/com/dotcms/experiments/business/ExperimentsAPIImpl.java

```diff
@@ -0,0 +1,140 @@
+import com.dotcms.experiments.model.Experiment;
+import com.dotcms.util.DotPreconditions;
+import com.dotmarketing.business.APILocator;
+import com.dotmarketing.business.DotStateException;
+import com.dotmarketing.business.FactoryLocator;
+import com.dotmarketing.business.PermissionAPI;
+import com.dotmarketing.business.PermissionLevel;
+import com.dotmarketing.exception.DotDataException;
+import com.dotmarketing.exception.DotSecurityException;
+import com.dotmarketing.portlets.contentlet.business.ContentletAPI;
+import com.dotmarketing.portlets.contentlet.model.Contentlet;
+import com.dotmarketing.util.Logger;
+import com.dotmarketing.util.UUIDGenerator;
+import com.dotmarketing.util.UtilMethods;
+import com.liferay.portal.model.User;
+import java.time.Instant;
+import java.time.LocalDateTime;
+import java.util.Optional;
+
+public class ExperimentsAPIImpl implements ExperimentsAPI {
+
+    final ExperimentsFactory factory = FactoryLocator.getExperimentsFactory();
+    final PermissionAPI permissionAPI = APILocator.getPermissionAPI();
+    final ContentletAPI contentletAPI = APILocator.getContentletAPI();
+
+    @Override
+    @WrapInTransaction
+    public Experiment save(final Experiment experiment, final User user) throws
+            DotSecurityException, DotDataException {
+
+        final Contentlet pageAsContent = contentletAPI
+                .findContentletByIdentifierAnyLanguage(experiment.pageId(), false);
+
+        DotPreconditions.isTrue(pageAsContent!=null
+                && UtilMethods.isSet(pageAsContent.getIdentifier()),
+                DotStateException.class, ()->"Invalid Page provided");
+
+        if(!permissionAPI.doesUserHavePermission(pageAsContent, PermissionLevel.EDIT.getType(), user)) {
+            Logger.error(this, "You don't have permission to save the Experiment.");
+            throw new DotSecurityException("You don't have permission to save the Experiment.");
+        }
+
+        Experiment.Builder builder = Experiment.builder().from(experiment);
+
+        if(experiment.id().isEmpty()) {
+            builder.id(UUIDGenerator.generateUuid());
+        }
+
+        builder.modDate(Instant.now());
+        builder.lastModifiedBy(user.getUserId());
+
+        final Experiment experimentToSave = builder.build();
+
+        return factory.save(experimentToSave);
+    }
+
+    @Override
+    public Optional<Experiment> find(final String id, final User user)
+            throws DotDataException, DotSecurityException {
+        if(!UtilMethods.isSet(id)) {
+            return Optional.empty();
+        }
+
+        final Optional<Experiment> experiment =  factory.find(id);
+
+        if(experiment.isPresent()) {
+            final Contentlet pageAsContent = contentletAPI
+                    .findContentletByIdentifierAnyLanguage(experiment.get().pageId(), false);
+
+            if (!permissionAPI.doesUserHavePermission(pageAsContent, PermissionLevel.EDIT.getType(),
```

Comment:
since you have the id, you can safe an api call by using a Permission Proxy

Reply:
Done